src/main/java/io/castle/client/api/CastleApiImpl.java

```diff
@@ -0,0 +1,141 @@
+    }
+
+    private CastleApiImpl(HttpServletRequest request, boolean doNotTrack, CastleSdkInternalConfiguration configuration, JsonObject contextJson) {
+        this.request = request;
+        this.doNotTrack = doNotTrack;
+        this.configuration = configuration;
+        this.contextJson = contextJson;
+    }
+
+    private CastleContext buildContext() {
+        CastleContextBuilder builder = new CastleContextBuilder(configuration.getConfiguration());
+        CastleContext context = builder
+                .fromHttpServletRequest(this.request)
+                .build();
+        return context;
+    }
+
+    @Override
+    public CastleApi mergeContext(Object additionalContext) {
+        JsonObject contextToMerge = configuration.getModel().getGson().toJsonTree(additionalContext).getAsJsonObject();
+        JsonObject mergedContext = new ContextMerge().merge(this.contextJson, contextToMerge);
+        return new CastleApiImpl(request, doNotTrack, configuration, mergedContext);
+    }
+
+    @Override
+    public CastleApi doNotTrack(boolean doNotTrack) {
+        return new CastleApiImpl(request, doNotTrack, configuration);
+    }
+
+    @Override
+    public AuthenticateAction authenticate(String event, String userId) {
+        return authenticate(event, userId, null);
+    }
+
+    @Override
+    public AuthenticateAction authenticate(String event, String userId, Object properties) {
+        RestApi restApi = configuration.getRestApiFactory().buildBackend();
+        JsonElement propertiesJson = null;
+        if (properties != null) {
+            propertiesJson = configuration.getModel().getGson().toJsonTree(properties);
+        }
+        return restApi.sendAuthenticateSync(event, userId, contextJson, propertiesJson);
+    }
+
+    @Override
+    public void authenticateAsync(String event, String userId, @Nullable Object properties, AsyncCallbackHandler<AuthenticateAction> asyncCallbackHandler) {
+        RestApi restApi = configuration.getRestApiFactory().buildBackend();
+        JsonElement propertiesJson = null;
+        if (properties != null) {
+            propertiesJson = configuration.getModel().getGson().toJsonTree(properties);
+        }
+        restApi.sendAuthenticateAsync(event, userId, contextJson, propertiesJson, asyncCallbackHandler);
+
+    }
+
+    @Override
+    public void authenticateAsync(String event, String userId, AsyncCallbackHandler<AuthenticateAction> asyncCallbackHandler) {
+        authenticateAsync(event, userId, null, asyncCallbackHandler);
+    }
+
+    @Override
+    public void track(String event) {
+        track(event, null, null, null);
+    }
+
+    @Override
+    public void track(String event, String userId) {
+        track(event, userId, null, null);
+    }
+
+    @Override
+    public void track(String event, String userId, Object properties) {
+        track(event, userId, properties, null);
```

Comment:
should support optional context see: https://paper.dropbox.com/doc/Server-SDK-specs-5WNJdpiGefPk2I1aAtWET#:h2=Method-call this applies for `track, identify, authenticate`

Reply:
The API is a fluend interface where several context can be added.
See the test on CastleMergeHttpTest.

To merge additional context the api is used as:
 sdk.onRequest(request).mergeContext(customExtraContext).identify(id, true);

This allows to merge more that one context. The sygnature is Object, so any data class can be used.

Reply:
ok thanks for clarification